internal/compiler-bridge/src/main/scala/xsbt/Dependency.scala

```diff
@@ -114,8 +114,9 @@ final class Dependency(val global: CallbackGlobal) extends LocateClassFile with
             // The dependency comes from a JAR
             for {
               zip <- zipEntry.underlyingSource
-              classFile <- Option(zip.file)
-            } binaryDependency(classFile, binaryClassName)
+              jarFile <- Option(zip.file)
+              if !jarFile.isDirectory // workaround for JDK9 and Scala 2.10/2.11, see https://github.com/sbt/sbt/pull/3701
+            } binaryDependency(jarFile, binaryClassName)
```

Comment:
This is the only change that fixes an observable bug. I didn't figure out why the bug in SBT 0.13 fixed by `-Dscala.ext.dirs` was not necessary here.

Reply:
Can you elaborate why Scalac reports the underlying source to be a directory under JDK9? I'm curious.

Reply:
To integrate the JEP-220 support into Scala 2.10/2.11, I made the `jrt://` virtual file system appear as a `ZipFile`. This was done by subclassing, but the superclass had inadvertently overriden `override val underlyingSource = Some(jarFile)`. I could not override this to return `None` without making a binary compatible change to the method I was overriding (widening to `Option`).

So I return `Some($JAVA_HOME)` instead, but that messed up Zinc.

Rocks and hard places, all around.

Reply:
Great to know this, thanks for the explanation.